backend/src/stwo/prover.rs

```diff
@@ -1,32 +1,144 @@
+
+use stwo_prover::constraint_framework::TraceLocationAllocator;
+use stwo_prover::core::prover::StarkProof;
+
 use powdr_number::FieldElement;
+use stwo_prover::core::air::Component;
+use stwo_prover::core::air::ComponentProver;
+use stwo_prover::core::backend::Backend;
+use stwo_prover::core::backend::BackendForChannel;
+use stwo_prover::core::channel::Channel;
+use stwo_prover::core::channel::MerkleChannel;
+use stwo_prover::core::fields::m31::M31;
+use stwo_prover::core::fri::FriConfig;
+use stwo_prover::core::pcs::{CommitmentSchemeProver, CommitmentSchemeVerifier, PcsConfig};
+use stwo_prover::core::poly::circle::CanonicCoset;
+
+const FRI_LOG_BLOWUP: usize = 1;
+const FRI_NUM_QUERIES: usize = 100;
+const FRI_PROOF_OF_WORK_BITS: usize = 16;
+const LOG_LAST_LAYER_DEGREE_BOUND: usize = 0;
 
 #[allow(unused_variables)]
-pub struct StwoProver<F> {
-    _analyzed: Arc<Analyzed<F>>,
-    _fixed: Arc<Vec<(String, Vec<F>)>>,
+pub struct StwoProver<T, B: Backend + Send, MC: MerkleChannel, C: Channel> {
+    pub analyzed: Arc<Analyzed<T>>,
+    _fixed: Arc<Vec<(String, Vec<T>)>>,
     /// Proving key placeholder
     _proving_key: Option<()>,
     /// Verifying key placeholder
     _verifying_key: Option<()>,
+    pub pcs_config: PcsConfig,
+    _channel_marker: PhantomData<C>,
+    _channel_marker_backend: PhantomData<B>,
+    _channel_marker_merkelchannel: PhantomData<MC>,
 }
 
-impl<F: FieldElement> StwoProver<F> {
+impl<'a, F: FieldElement, B: Backend + Send, MC: MerkleChannel + Send, C: Channel + Send>
+    StwoProver<F, B, MC, C>
+where
+    B: Backend + BackendForChannel<MC>, // Ensure B implements BackendForChannel<MC>
+    MC: MerkleChannel,
+    C: Channel,
+    MC::H: DeserializeOwned + Serialize,
+    PowdrComponent<'a, F>: ComponentProver<B>,
+{
     #[allow(dead_code)]
     #[allow(unused_variables)]
     pub fn new(
-        _analyzed: Arc<Analyzed<F>>,
+        analyzed: Arc<Analyzed<F>>,
         _fixed: Arc<Vec<(String, Vec<F>)>>,
-        setup: Option<&mut dyn io::Read>,
+        //  setup: Option<&mut dyn io::Read>,
     ) -> Result<Self, io::Error> {
+        let config = PcsConfig {
+            pow_bits: FRI_PROOF_OF_WORK_BITS as u32,
+            fri_config: FriConfig::new(
+                LOG_LAST_LAYER_DEGREE_BOUND as u32,
+                FRI_LOG_BLOWUP as u32,
+                FRI_NUM_QUERIES,
+            ),
+        };
+
         Ok(Self {
-            _analyzed,
+            analyzed,
             _fixed,
             _proving_key: None,
             _verifying_key: None,
+            _channel_marker_backend: PhantomData,
+            _channel_marker_merkelchannel: PhantomData,
+            pcs_config: config,
+            _channel_marker: PhantomData,
         })
     }
+    pub fn prove(&self, witness: &[(String, Vec<F>)]) -> Result<Vec<u8>, String> {
+        // twiddles are used for FFT, it is computed in a bigger group than the eval domain.
+        // eval domain half coset G_{2n} + <G_{n/2}>
```

Comment:
```suggestion
        // the eval domain is the half coset G_{2n} + <G_{n/2}>
```
?

Reply:
Also what does `<...>` mean here?

Reply:
short answer: when creating twiddles, all the examples given in stwo, has this +1 check[ here](https://github.com/starkware-libs/stwo/blob/64270dc51420a6887e7c0695c71584949711dcb2/crates/prover/src/examples/wide_fibonacci/mod.rs#L176), some related comments in[ stwo ](https://github.com/starkware-libs/stwo/blob/64270dc51420a6887e7c0695c71584949711dcb2/crates/prover/src/core/poly/twiddles.rs#L4) and [here](https://github.com/starkware-libs/stwo/blob/64270dc51420a6887e7c0695c71584949711dcb2/crates/prover/src/core/circle.rs#L332)

long answer:
<...> means the generator of the subgroup of the circle group, <G_{n}> generates a group of size 2^n.  in coset (eval domain is a coset) context, it is referred as step size, G_{2n} is the offset of the coset. 
say a trace domain of size 4 in normal STARK, the cyclic group can be written as <G_2>, eval domain is a coset, then it can be written as G_3 + <G_2>, or G_4 + <G_2>, as long as the offset is not in <G_2>
In circle STARK, trace domain itself is a coset, for coset of size 4,  it can be written as G_4+<G_2> as well, but it can also be written as G_4+<G_1> U G_4^{-1}+<G_1>, this is the twin coset/ standard position coset, mentioned often in circle STARK tutorials. The half coset is  G_4+<G_1>, its size is 2. 
eval coset will be G_8+<G_2> U G_8^{-1}+<G_2> of size 8
twiddles are used for FFT, and they will be computed on a half coset G_5+<G_3> of size 8, it comes from the coset G_5+<G_3> U G_5^{-1}+<G_3> of size 16.   one log size bigger than eval domain  
I am writing some blogs to give some explanations for this, including FFT in circle domain. TBD...